pkg/scheduler/controller/shoot/reconciler.go

```diff
@@ -217,12 +219,27 @@ func filterSeedsMatchingProviders(cloudProfile *gardencorev1beta1.CloudProfile,
 	return matchingSeeds, nil
 }
 
+// filterHighAvailabilitySeeds filters seeds based on their high-availability configurations
+func filterHighAvailabilitySeeds(seedList []gardencorev1beta1.Seed, shoot *gardencorev1beta1.Shoot) ([]gardencorev1beta1.Seed, error) {
```

Comment:
Is it possible to add a env test to covert this case not only with unit test but also env test test case?

https://github.com/gardener/gardener/blob/52e9635b0bd2009a92d7801e976e4420f6bba3bc/test/integration/scheduler/shoot/scheduler_suite_test.go#L41

Reply:
This comment was a little bit "hijacked". 😅 

My initial suggest was to add new test cases to the scheduler env test that verify:
- HA Shoot with `node` failureTolerance type will fail to be scheduled if there are no HA Seeds with failureTolerance type `node` or `zone`
- HA Shoot with `node` failureTolerance type will succeed to be scheduled when there is a HA Seed with failureTolerance type `node`

Reply:
There are some new env tests for the scenario. I reworked the existing env tests for high availability too. They did not work properly, because the `seed` and `shoot` objects were changed after they have been created on the api-server.

Additionally, I added some tests for HA and minimal distance scheduler and general minimal distance tests.

I might have found an error in the minimal distance calculation. In my opinion [this test](https://github.com/oliver-goetz/gardener/blob/5c3b5f1128429aea2950770f68f6e781f530bc5e/test/integration/scheduler/shoot/scheduler_test.go#L241-L256) should fail, because `eu-west-1` is closer to `us-east-2` than `eu-east-2`. This would be something for a different PR though.

Reply:
> I might have found an error in the minimal distance calculation. In my opinion [this test](https://github.com/oliver-goetz/gardener/blob/5c3b5f1128429aea2950770f68f6e781f530bc5e/test/integration/scheduler/shoot/scheduler_test.go#L241-L256) should fail, because eu-west-1 is closer to us-east-2 than eu-east-2. This would be something for a different PR though.

(Probably) makes sense. Reading the [implementation details](https://github.com/gardener/gardener/blob/master/docs/concepts/scheduler.md#strategies) though, I'd say that this is expected. You can open an issue and link @mandelsoft, who can provider provide more input 🙂